File: packages/evals/src/quality.ts
import type { Judge, JudgeSpecInput, JudgeVerdict } from "./judge.js";

/**
 * L1 Spec quality regression harness. Runs alongside runGolden (whether the output Spec matches the expectation).
 * Whereas runGolden checks "identity" (a match after variation normalization), this regression-checks the
 * L1 quality rubric's score via judge.judgeSpec (whether the lower bound of "how good" is upheld).
 * It can run deterministically with a judge that has FakeLlm / FixtureLlm injected (live verification of the LLM path is record/replay).
 *
 * Positioning: this harness provides the mechanism (case → judgeSpec → lower-bound gate → report) **up front**.
 * Defining the actual quality cases and wiring up FixtureLlm (record/replay) is the responsibility of the
 * consumer (the product's regression suite); here we provide only the minimal, deterministically running executor.
 */
export interface QualityCase {
  name: string;
  input: JudgeSpecInput;
  /**
   * Lower bound of the expected score. When specified, the pass condition is verdict.score >= minScore (a regression threshold gate).
   * When unspecified, verdict.pass (based on the judge's passScore) is used directly.
   */
  minScore?: number;
}

export interface QualityCaseResult {
  name: string;
  pass: boolean;
  score: number;
  verdict: JudgeVerdict;
  /**
   * The case's execution time (wall clock). **Not deterministic**: pass / score / verdict are deterministic
   * with FakeLlm/FixtureLlm injected, but durationMs is a measured value and varies per run (do not include it in regression comparisons or snapshots).
   */
  durationMs: number;
}

export interface QualityReport {
  pass: boolean;
  cases: QualityCaseResult[];
}

/** Runs the L1 quality cases through judge.judgeSpec for regression verification. */
export async function runQuality(cases: QualityCase[], judge: Judge): Promise<QualityReport> {
  const results: QualityCaseResult[] = [];
  for (const c of cases) {
    const startedAt = Date.now();
    const verdict = await judge.judgeSpec(c.input);
    const pass = c.minScore != null ? verdict.score >= c.minScore : verdict.pass;
    results.push({
      name: c.name,
      pass,
      score: verdict.score,
      verdict,
      durationMs: Date.now() - startedAt,
    });
  }
  return { pass: results.every((r) => r.pass), cases: results };
}

File: python/kohaku/src/kohaku/evals/quality.py
"""L1 Spec quality-regression harness (port of TS packages/evals/src/quality.ts; B7).

Runs alongside run_golden (whether the output Spec matches the expectation). Whereas run_golden checks
"identity (a variance-normalized match)", this regression-checks the L1 quality rubric score via
judge.judge_spec (whether it holds the "quality" lower bound). It can run deterministically with a judge
into which FakeLlm / FixtureLlm has been injected.

Positioning: this harness provides the mechanism up front (case → judge_spec → lower-bound gate → report).
Defining the actual quality cases and wiring FixtureLlm (record/replay) is the consumer's responsibility
(the product's regression suite); here we provide only the minimal, deterministically-running executor.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

from .judge import Judge, JudgeSpecInput, JudgeVerdict


@dataclass(frozen=True)
class QualityCase:
    name: str
    input: JudgeSpecInput
    min_score: float | None = None
    """The lower bound of the expected score. When given, verdict.score >= min_score is the pass condition
    (the regression threshold gate). When omitted, verdict.pass_ (based on the judge's pass_score) is used as-is."""


@dataclass(frozen=True)
class QualityCaseResult:
    name: str
    pass_: bool
    score: float
    verdict: JudgeVerdict
    duration_ms: float
    """The case's execution time (wall clock). Out of scope for determinism: pass_ / score / verdict are
    deterministic with FakeLlm/FixtureLlm injection, but duration_ms is a measured value that varies per run
    (not included in the regression comparison)."""


@dataclass(frozen=True)
class QualityReport:
    pass_: bool
    cases: list[QualityCaseResult]


async def run_quality(cases: list[QualityCase], judge: Judge) -> QualityReport:
    """Run the L1 quality cases through judge.judge_spec and regression-verify them."""
    results: list[QualityCaseResult] = []
    for c in cases:
        started_at = time.monotonic()
        verdict = await judge.judge_spec(c.input)
        passed = verdict.score >= c.min_score if c.min_score is not None else verdict.pass_
        results.append(
            QualityCaseResult(
                name=c.name,
                pass_=passed,
                score=verdict.score,
                verdict=verdict,
                duration_ms=(time.monotonic() - started_at) * 1000,
            )
        )
    return QualityReport(pass_=all(r.pass_ for r in results), cases=results)
